### PR TITLE
Jay - PERMISSION: Edit Time Entry Description Permission

### PR DESCRIPTION
**Pull request by Alforoan**
# Description
This PR enables the user with the Edit Time Entry Description permission to be able to access any user's time entry description and edit it and update successfully.

## Related PRS (if any):
This frontend PR is related to the [#712](https://github.com/OneCommunityGlobal/HGNRest/pull/712) backend PR.

…

## Main changes explained:
- Added a new permission label for Edit Time Entry Description permission.
- Added conditional logic so that the user with the permission can view the edit button and edit the time log description.

## How to test (please watch the video):
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
4. log in with an owner account
5. go to Other Links -> Permissions Management -> Manage User Permissions -> Choose a user without Edit Time Entry Description permission under Timelog Management, Edit Time Entries, Edit Timelog Description -> click on Add  -> Scroll down then submit
6. verify that the user who was given the permission is now able to see the the edit button on another user's dashboard for time log entries.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/3dc44aab-1f81-4f26-8c97-09f82afe301c

## Note:
**Unmute the video**

### CODE REVIEW

**Review by X1aoZhang99 — DISMISSED** [review]
Hi Jay,

I've tested your PR, and all functionalities worked as intended.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/1daf8267-d581-4899-a49f-dc14460fef72

**Review by abdel-lall — DISMISSED** [review]
Works as described. Good job!
Screenshot 2024-01-27 185905
Screenshot 2024-01-27 190018

Screenshot 2024-01-27 185940

**Review by Renanluizssx — DISMISSED** [review]
Hi @Alforoan, i tested your application and it's correct, see my review below:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91035340/5297c326-6afb-4c31-b87c-2ba38ea16b42

**Review by Nahiyan-16 — DISMISSED** [review]
## Works fine
Hi Jay, just tested your PR and everything seems to work fine. I logged in as an owner and updated permission of another account to be able to edit the description. I made sure that when it was toggled off you weren't able to edit it and when you toggle it on that you are able to. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/71025942/82539da9-e7d5-49ce-9c8a-2f5404aac5cd

**Review by Pandani07 — DISMISSED** [review]
Tested this PR and everything works as intended. I added the edit time entry permission for my admin account and I verified if I could change the time log description. The edited description is saved after the update as well. Nice work

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/f7767035-df98-4d3a-838f-98678f7e5283

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 1]
```diff
@@ -690,7 +693,7 @@ const TimeEntryForm = props => {
                   name="isTangible"
                   checked={formValues.isTangible}
                   onChange={handleInputChange}
-                  disabled={!canEditTimeEntry && !initialIsTangible}
+                  disabled={(!canEditTimeEntry && !initialIsTangible || canEditTimeEntryDescription)}
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:696]:
```suggestion
                  disabled={(!canEditTimeEntry && !initialIsTangible && !canEditTimeEntryDescription)}
```
Does this need to be inversed?

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 2]
```diff
@@ -195,7 +196,7 @@ const TimeEntryForm = props => {
     const today = moment().tz('America/Los_Angeles');
     const isDateValid = date.isValid();
     // Administrator/Owner can add time entries for any dates, and other roles can only edit their own time entry in the same day.
-    const isUserAuthorized = (canEditTimeEntry && canPutUserProfileImportantInfo) || !edit || today.diff(date, 'days') === 0
+    const isUserAuthorized = (canEditTimeEntry && canPutUserProfileImportantInfo) || canEditTimeEntryDescription || !edit || today.diff(date, 'days') === 0
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:199]:
```suggestion
    const isUserAuthorized = (canEditTimeEntry && canPutUserProfileImportantInfo) || !edit || today.diff(date, 'days') === 0
```
isUserAuthorized looks like it's only about changing dates of time entries, not descriptions?

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 3]
```diff
@@ -655,6 +657,7 @@ const TimeEntryForm = props => {
             <FormGroup>
               <Label for="notes">Notes</Label>
               <Editor
+                disabled={!(canEditTimeEntry || canEditTimeEntryDescription) }
```
Comment (nathanah) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:660]:
```suggestion
                disabled={!canEditTimeEntryDescription}
```
`editTimeEntry` should be replaced by more granular permissions (`editTimeEntryDescription` in this case)

**Review by nathanah — CHANGES_REQUESTED** [review]

**Review by mohabbasd — COMMENTED** [review]
Hello, I tested your PR and I added the permission to my test account from the owner account but unfortunately I cannot see any other user other than myself and even if I get other user's information I am not able to see the time entry.

Screenshot 2024-02-16 at 1 14 32 AM
Screenshot 2024-02-16 at 1 19 06 AM

**Review by XiaohanMeng — DISMISSED** [review]
Hi, I tested this PR using the owner, volunteer, and admin roles, and everything worked as intended! After the owner granted permissions, volunteers could edit their own Time Entry Descriptions, while admins were able to edit the Time Entry Descriptions of other users. Please see the reference video.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113215317/1ae91b78-765d-4bba-9619-a9c6141293a5

**Review by Coops023 — DISMISSED** [review]
Hi Jay,

i reviewed your PR and tested functionality of giving permission to a user without time entry editing permission and it is working as described by your instructions. well done!

PR-1859-edited

PR-1859-editform

**Review by Jingyii800 — DISMISSED** [review]
Hi, I have tested this feature, and it worked well. As an owner role, I can give one user permission to let it change the others' time log description. Please check this video for more details. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/112589476/de472d98-ac43-4623-9b4d-cad1097fdc8a

**Review by ClemarNunes — DISMISSED** [review]
I tested your PR and worked as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/134466562/7d0ed21f-1b41-4257-be28-07d48732aebc

**Comment by Weiyao-Li** [comment]
I've tested this feature, and it performed well. As the owner, I added the time log edition permission to volunteer. And volunteer can be 
Screenshot 2024-02-23 at 16 42 33
Screenshot 2024-02-23 at 16 39 24
Screenshot 2024-02-23 at 16 39 17
Screenshot 2024-02-23 at 16 23 35
able to see and edit its own time log.

**Review by SRamen1999 — COMMENTED** [review]
All functionalities worked as intended. Used owner account and gave volunteer account timelog description permission and was able to edit past timelog descriptions.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/2a4e17d8-962c-4753-81fc-dd93de635d97

**Comment by rajnada** [comment]
Hii,
I have followed the steps and review the functionality it's working fine for me.

Give permission to User
1859B

Initial Time Entry
1859B2

Edit Time Entry
1859A

**Review by hetvi11110 — CHANGES_REQUESTED** [review]
I attempted to log in as indicated here, but I'm facing a white screen on the user dashboard page. Here's a screenshot of the console error.

image

**Review by iven-yao — CHANGES_REQUESTED** [review]
without the permission, users cannot even log in their own time entries, project/task drop down selection and notes input area are both disabled. 
image
image

File: src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx [hunk 4]
```diff
@@ -663,7 +667,9 @@ const TimeEntryForm = props => {
                     placeholder="Hours"
                     value={formValues.hours}
                     onChange={handleInputChange}
-                    disabled={!canChangeTime}
+                    disabled={!((canEditTimeEntry && !fromTimer) || canChangeTime)}
```
Comment (kylepham) [inline src/components/Timelog/TimeEntryForm/TimeEntryForm.jsx:670]: `fromTimer` is not previously defined.

**Review by kylepham — CHANGES_REQUESTED** [review]
There is an error of undefined variable `fromTimer` that caused the app to display blank screen on startup. 
Screenshot 2024-04-01 at 3 05 55 PM (2)


I temporarily put `const fromTimer = false` in the modal component and everything worked as intended.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/16fbe89d-7c09-4e07-a9f0-a744a64c560c

**Pull request closed without merging**